compiler/pipes/check-color.h

```diff
@@ -0,0 +1,373 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2020 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#pragma once
+
+#include <utility>
+
+#include "common/algorithms/contains.h"
+#include "common/termformat/termformat.h"
+#include "compiler/compiler-core.h"
+#include "compiler/function-colors.h"
+#include "compiler/function-pass.h"
+
+namespace fp = function_palette;
+
+// Callstack class describes a static array for storing functions
+// while checking a function. An array is allocated on the stack,
+// which avoids heap allocations, which allows less memory to be used.
+class Callstack {
+  // Size constant describes the maximum depth of color checking,
+  // as well as the maximum number of functions in the callstack.
+  const static auto Size = 50;
+
+private:
+  FunctionPtr data[Size]{};
+  size_t count{0};
+
+public:
+  void add(const FunctionPtr& func) {
+    if (this->count >= Size) {
+      return;
+    }
+    this->data[this->count] = func;
+    this->count++;
+  }
+
+  bool contains(const FunctionPtr& func) const {
+    for (int i = 0; i < this->count; ++i) {
+      if (this->data[i] == func) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  const FunctionPtr& front() const {
+    return this->data[0];
+  }
+
+  const FunctionPtr& back() const {
+    return this->data[this->count - 1];
+  }
+
+  size_t size() const {
+    return this->count;
+  }
+
+  bool full() const {
+    return this->count == Size;
+  }
+
+  const FunctionPtr& operator[](size_t index) const {
+    if (index >= Size) {
+      return this->data[Size - 1];
+    }
+    return this->data[index];
+  }
+};
+
+class CheckColorPass final : public FunctionPassBase {
+private:
+  fp::RuleTree tree;
+
+public:
+  CheckColorPass() {
+    tree = G->get_function_palette_tree();
+  }
+
+  void check() const {
+    Callstack stacktrace{};
+    this->check_step(std::move(stacktrace), this->tree.root(), this->current_function);
+  }
+
+  void on_start() final {
+    this->check();
+  }
+
+  void on_finish() override {}
+
+  std::string get_description() override {
+    return "Check function color";
+  }
+
+  bool check_function(FunctionPtr function) const override {
+    return !function->is_extern();
+  }
+
+private:
+  // check_func_caller is a function that checks functions above the given in callstack.
+  void check_func_caller(const Callstack &stacktrace, fp::Node *rule, const FunctionPtr &func) const {
+    // If the current rule describes an error.
```

Comment:
We don't have a good document that describes the comments style, but there is a consensus that small inline comments inside a function/method should not start from a capital letter and are preferred not to have a period that ends them.

Take a look at the recently committed code to infer the comment rules. They're not strict, but it's easier if we try to keep that style in sync.

So:
```diff
- // If the current rule describes an error.
+ // if the current rule describes an error
```

Reply:
Thanks, corrected.